File: solution/c#/Day23/Day23/Trips/TripService.cs
using Day23.Exception;
using Day23.Users;
using LanguageExt;
using static LanguageExt.Prelude;
using Seq = LanguageExt.Seq;

namespace Day23.Trips
{
    public class TripService(ITripRepository repository)
    {
        private static readonly Seq<Trip> NoTrips = Seq.empty<Trip>();

        public Try<Seq<Trip>> RetrieveFriendTrips(User user, User? loggedUser)
            => CheckUser(loggedUser, () => RetrieveFriendsSafely(user, loggedUser!));

        private static Try<Seq<Trip>> CheckUser(
            User? loggedUser,
            Func<Seq<Trip>> continueWith)
            => loggedUser != null
                ? Try(continueWith())
                : Try<Seq<Trip>>(new UserNotLoggedInException());

        private Seq<Trip> RetrieveFriendsSafely(User user, User loggedUser)
            => user.IsFriendWith(loggedUser)
                ? FindTripsByUser(user)
                : NoTrips;

        private Seq<Trip> FindTripsByUser(User user) => repository.FindTripsByUser(user);
    }
}

File: solution/ts/day23/src/trip/TripService.ts
import {TripRepository} from "./TripRepository";
import {User} from "../user/User";
import Trip from "./Trip";
import * as E from "fp-ts/Either";
import UserNotLoggedInException from "../exception/UserNotLoggedInException";

const noTrips: Trip[] = [];

export class TripService {
    constructor(private readonly repository: TripRepository) {
    }

    retrieveFriendTrips = (user: User, loggedUser: User): E.Either<UserNotLoggedInException, Trip[]> =>
        this.checkUser(loggedUser)
            ? E.right(this.retrieveFriendsSafely(user, loggedUser))
            : E.left(new UserNotLoggedInException());
    private retrieveFriendsSafely = (user: User, loggedUser: User): Trip[] =>
        user.isFriendWith(loggedUser)
            ? this.findTripsByUser(user) :
            noTrips;
    private checkUser = (loggedUser: User): boolean => loggedUser != null;
    private findTripsByUser = (user: User): Trip[] => this.repository.findTripsByUser(user);
}